### PR TITLE
Fix bug in slow tokenizer conversion, make it a lot faster

### LINKED ISSUE #24233: Auto-Converted Fast Tokenizer Producing Incorrect Results
### System Info

- `transformers` version: 4.30.1
- Platform: Linux-5.15.107+-x86_64-with-glibc2.31
- Python version: 3.10.12
- Huggingface_hub version: 0.15.1
- Safetensors version: 0.3.1
- PyTorch version (GPU?): 2.0.1+cu118 (False)
- Tensorflow version (GPU?): 2.12.0 (False)
- Flax version (CPU?/GPU?/TPU?): 0.6.9 (cpu)
- Jax version: 0.4.10
- JaxLib version: 0.4.10
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

The auto-converted fast tokenizer for the LLaMA model sometimes does not produce the same tokenization results as the original sentence piece tokenizer. This is affecting the OpenLLaMA models. Here's the code to reproduce it:

```python
from transformers import AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained('openlm-research/open_llama_7b', use_fast=False)
fast_tokenizer = AutoTokenizer.from_pretrained('openlm-research/open_llama_7b')

text = 'thermal'
print(tokenizer.encode(text))
print(fast_tokenizer.encode(text))
```

The code produces the following output:

```
[1, 14412]
[1, 31822, 496, 12719]
```

### Expected behavior

The auto-converted fast tokenizer should produce the exact same tokens as the original sentencepiece tokenizer.

### PR DESCRIPTION
**Pull request by stephantul**
# What does this PR do?

The slow tokenizer conversion currently has a bug where merges with a score of 0 do not get used due to an erroneous check. The check simply tested truthiness, but was actually looking for a `None`. During fixing, I noticed that the code was also slow, so I made it a lot faster.

Fixes #24233

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker

### CODE REVIEW

**Comment by stephantul** [comment]
Speed info: the new implementation takes 70 ms, the old one took 123 seconds for the `openlm-research/open_llama_7b` tokenizer mentioned in the issue.

**Review by amyeroberts — APPROVED** [review]
Really nice fix and improvement - thanks for working on this ❤️ 

Logic all looks good to me. There's a test that's failing, but it's decorated with `@is_flaky` so shouldn't be preventing CI being green here. @ydshieh any insights into what might be happening?

**Comment by ydshieh** [comment]
> Really nice fix and improvement - thanks for working on this ❤️
> 
> Logic all looks good to me. There's a test that's failing, but it's decorated with `@is_flaky` so shouldn't be preventing CI being green here. @ydshieh any insights into what might be happening?

@amyeroberts 

`is_flacky()` won't keep the test green 100%. It just runs the test a few times (default `5`) 😿 . The failing is still expected but less frequently.

File: src/transformers/convert_slow_tokenizer.py [hunk 1]
```diff
@@ -54,12 +54,15 @@ def extract(self, vocab_scores=None) -> Tuple[Dict[str, int], List[Tuple]]:
...
-                piece_score = vocab_scores.get(merge, None)
-                if piece_score:
-                    merges += [(piece_l, piece_r, piece_score)]
+        for merge, piece_score in vocab_scores.items():
+            local = []
+            for index in range(1, len(merge)):
+                piece_l, piece_r = merge[:index], merge[index:]
+                if piece_l in vocab and piece_r in vocab:
+                    local.append((piece_l, piece_r, piece_score))
+            local = sorted(local, key=lambda x: (vocab[x[0]], vocab[[x[1]]]))
```
Comment (ArthurZucker) [inline src/transformers/convert_slow_tokenizer.py:63]:
```suggestion
            local = sorted(local, key=lambda x: (vocab[x[0]], vocab[x[1]]))
```
seems to fix the conversion for me 😉

**Review by ArthurZucker — COMMENTED** [review]
LGTM, but running `RUN_SLOW=1 RUN_TOKENIZER_INTEGRATION=1 pytest tests/models/llama/test_tokenization_llama.py ` (the code was changed for llama specifically) crashes. 
This probably happens when sorting `local = sorted(local, key=lambda x: (vocab[x[0]], vocab[[x[1]]]))`, left a suggestion. 

Also it seems that the new serialization differs from the old one, that's probably just the special tokens that are no longer normalized by default

**Comment by stephantul** [comment]
Sorry for the weird syntax error. I forgot to re-run tests after the second commit

**Review by ArthurZucker — APPROVED** [review]
Thanks a lot 🤗
